FR #1059: Expand Normative Requirements Guidelines with Core Authoring Rules and Attribute Requirements (#1967)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: guidelines/normative-requirements-guidelines.md

# Normative Requirements Guidelines

This section defines guidelines for authoring normative requirements in the FOCUS specification. These guidelines define **how** to write normative requirements to ensure clarity, consistency, and testability. It does not define the requirements themselves (the "what") but concentrates on their **structure, subjects, and verifiability**.

The guidelines cover normative requirements applicable to:

* **FOCUS Datasets** — the primary containers of structured data as defined in FOCUS.

> * **Columns** — individual columns within datasets (may contain nested objects and object properties, which can have additional normative rules)

Reviewer: ```suggestion
* **FOCUS Columns** — individual columns within FOCUS datasets (may contain nested objects and object properties, which can have additional normative rules)
```

### 4. State, Not Behavior

Normative requirements MUST describe a **verifiable state**, not an operational process or behavior.

Specifically:

* Process-oriented verbs such as *ensure*, *handle*, *support*, or *provide* MUST NOT be used.
* If a requirement refers to actor behavior, it MUST be reformulated as:
  * a constraint on the resulting dataset state, or
  * a constraint on a schema-defined artifact.

### 5. Use of BCP 14 Keywords

* Each normative bullet MUST contain exactly one BCP 14 keyword (MUST, SHOULD, MAY, MUST NOT, SHOULD NOT).
* A bullet containing more than one normative keyword MUST be split.

### 6. Splitting Requirements

A requirement MUST be split into multiple bullets if it:

* combines multiple obligations,
* combines a rule and an exception,
* mixes a definition with a constraint,
* applies different constraints to different subjects.

### 7. Composite Requirements

Composite (parent + nested) requirements MAY be used when strictly controlled and MUST adhere to the following requirements:

* Nested bullets MUST share the same condition if defined by the parent bullet.
* Nested bullets SHOULD NOT introduce a different subject.

> * When nested bullets introduce different subjects, all subjects SHOULD be of the same subject type (e.g., FOCUS * columns).

Author: ```suggestion
* When nested bullets introduce different subjects, all subjects SHOULD be of the same subject type (e.g., FOCUS * columns).

Composite requirements SHOULD be used when grouping improves readability and:

* multiple requirements share the same subject, or
* multiple requirements share the same business context (e.g., all requirements related to a specific business scenario or feature)

Flat parallel bullets SHOULD be preferred when ordering keyword is sufficient for clarity and readability.
```

### 4. State, Not Behavior

Normative requirements MUST describe a **verifiable state**, not an operational process or behavior.

Specifically:

* Process-oriented verbs such as *ensure*, *handle*, *support*, or *provide* MUST NOT be used.
* If a requirement refers to actor behavior, it MUST be reformulated as:
  * a constraint on the resulting dataset state, or
  * a constraint on a schema-defined artifact.

### 5. Use of BCP 14 Keywords

* Each normative bullet MUST contain exactly one BCP 14 keyword (MUST, SHOULD, MAY, MUST NOT, SHOULD NOT).
* A bullet containing more than one normative keyword MUST be split.

### 6. Splitting Requirements

A requirement MUST be split into multiple bullets if it:

* combines multiple obligations,
* combines a rule and an exception,
* mixes a definition with a constraint,
* applies different constraints to different subjects.

### 7. Composite Requirements

Composite (parent + nested) requirements MAY be used when strictly controlled.

They are allowed only when:

* the parent bullet introduces exactly one normative subject, and
* all nested bullets:
  * apply to that same subject,
  * do not introduce a new normative context.

> Nested bullets MUST NOT introduce a different subject.

Author: ```suggestion
Nested bullets MUST share the same condition if defined by the parent bullet.
Nested bullets SHOULD NOT introduce a different subject. 
When nested bullets introduce different subjects, all subjects SHOULD be of the same subject type (e.g., FOCUS columns).
```

# Normative Requirements Guidelines

> This section defines guidelines for authoring normative requirements in the FOCUS specification. These guidelines define **how** to write normative requirements to ensure clarity, consistency, and testability. It does not define the requirements themselves (the "what") but focuses on their **structure, subjects, and verifiability**.

Reviewer 2: ```suggestion
This section defines guidelines for authoring normative requirements in the FOCUS specification. These guidelines define **how** to write normative requirements to ensure clarity, consistency, and testability. It does not define the requirements themselves (the "what") but concentrates on their **structure, subjects, and verifiability**.
```

Removing focuses as to not confuse potential reader (FOCUS / focuses, get it?)

Reviewer 3: +1 to this. It's one of my primary ticks when I read or talk about the spec, but never thought how to articulate it and provide an alternative.

> This section defines guidelines for authoring normative requirements in the FOCUS specification. These guidelines define **how** normative requirements must be written to ensure clarity, consistency, and testability. It does not define the requirements themselves (the “what”) but focuses on their **structure, subjects, and verifiability**.

Reviewer 4: 🤖 [AI][Claude] Consider rephrasing to avoid "must" in non-normative context to prevent confusion with BCP-14 keywords.

```suggestion
This section defines guidelines for authoring normative requirements in the FOCUS specification. These guidelines define **how** to write normative requirements to ensure clarity, consistency, and testability. It does not define the requirements themselves (the "what") but focuses on their **structure, subjects, and verifiability**.
```

The guidelines cover normative requirements applicable to:

* **FOCUS Datasets** — the primary containers of structured data as defined in FOCUS.

> * **Columns** — individual columns within datasets (may contain nested Elements and ElementProperties, which can have additional normative rules))

Reviewer 4: 🤖 [AI][Claude] Typo: Extra closing parenthesis.

```suggestion
* **Columns** — individual columns within datasets (may contain nested Elements and ElementProperties, which can have additional normative rules)
```

* **FOCUS Dataset** — the primary dataset concept defined by the FOCUS specification.
* **Dataset Instance** — represents a concrete instantiation of a **FOCUS Dataset Instance**.
* **Dataset Artifact** — represents a physical or delivered form of a **FOCUS Dataset Instance Artifact**.

However, by design decision, the specification adopts the following normative conventions:

* **FOCUS Dataset is used as the canonical normative subject** for dataset-level requirements.
* Normative requirements are intentionally written against **FOCUS Dataset**, even when the constraint applies to:
  * a dataset specification,
  * a dataset instance, or
  * a dataset instance artifact.
* The intended level of application (specification vs. instance vs. artifact) is inferred from context rather than encoded in the normative subject.

This choice is intentional and overrides interpretations based solely on abstraction level.

## Core Normative Authoring Rules

### 1. Normative Requirement Structure

The recommended pattern for a normative requirement is:

``` markdown
<Subject (+qualifier)> + <BCP 14 Keyword> + <Verifiable State Descriptor> + <Object (+qualifier)> [+ Conditions]

> 

Reviewer 2: ```suggestion
```

### 3. Normative Subject

#### 3.1 Allowed Subjects

The normative subject MUST be a schema-level entity, such as:

* **FOCUS Dataset**, whereby use of:  
  * `FOCUS dataset` keyword represents any FOCUS dataset  
  * `FOCUS dataset` keyword with a qualifier represents a qualified subset of FOCUS datasets  
  * A single FOCUS dataset explicitly identified by `<FOCUS Dataset ID>` (e.g., `CostAndUsage`)

* **FOCUS Column**, whereby use of:  
  * `FOCUS column` keyword represents any FOCUS column  
  * `FOCUS column` keyword with a qualifier represents a qualified subset of FOCUS columns (e.g., `FOCUS column containing numeric values`)  
  * A single FOCUS column explicitly identified by `<FOCUS Column ID>` (e.g., `BilledCost`)

* **Custom Column**, whereby use of:  
  * `Custom column` keyword represents any custom column  
  * `Custom column` keyword with a qualifier represents a qualified subset of custom columns (e.g., `Custom column containing numeric values`)

* **Structural sub-elements within Columns** (objects, keys, key values):  

>   *Note: Do not use `object`, `key`, or `value` keywords alone. Always reference them in context, e.g.:*  

Reviewer 2: ```suggestion
  *Note: MUST NOT use `object`, `key`, or `value` keywords alone. Always reference them in context, e.g.:*  
```
Let's use BCP-14

  * `Object in Columns containing JsonObjectFormat values`  
  * `Key in Object in FOCUS/Custom column containing JsonObjectFormat values`  
  * `Key value in Object in FOCUS/Custom column containing key-value pairs`

The subject SHOULD be explicit and unambiguous.

**Exception:** A structural anchor requirement MAY use the ID of the enclosing schema construct as its subject, even if that construct is not otherwise an allowed normative subject (e.g., `<FOCUS Attribute  ID>`). This exception applies only to structural anchor requirements.

#### 3.2 Disallowed Subjects

The following MUST NOT be used as normative subjects:

* Actors (e.g. data generator, service provider, consumer)
* Processes or mechanisms (e.g. Delivery Handling, Correction Handling, etc.)

### 4. State, Not Behavior

Normative requirements MUST describe a **verifiable state**, not an operational process or behavior.

Specifically:

* Process-oriented verbs such as *ensure*, *handle*, *support*, or *provide* MUST NOT be used.

> * If a requirement refers to actor behavior, it MUST be reformulated as:

Reviewer 2: reformulated -> will this kick off an AI to do any rework?

Author: ```suggestion
* If a requirement refers to actor behavior, it MUST be expressed as:
```

Author: I here you. Updating the wording to “expressed as”. 
@ljadvey, please feel free to reopen the thread or propose alternative wording.

  * a constraint on the resulting dataset state, or
  * a constraint on a schema-defined artifact.

### 5. Use of BCP 14 Keywords

* Each normative bullet MUST contain exactly one BCP 14 keyword (MUST, SHOULD, MAY, MUST NOT, SHOULD NOT).
* A bullet containing more than one normative keyword MUST be split.

### 6. Splitting Requirements

A requirement MUST be split into multiple bullets if it:

* combines multiple obligations,
* combines a rule and an exception,
* mixes a definition with a constraint,
* applies different constraints to different subjects.

### 7. Composite Requirements

Composite (parent + nested) requirements MAY be used when strictly controlled.

They are allowed only when:

* the parent bullet introduces exactly one normative subject, and
* all nested bullets:

>   * apply to that same subject,

Reviewer 5: ```suggestion
* apply to that same subject,
```

Author: discarding the whole block in favor of [this](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1967/files#r2911518287) and resolving open thread

>   * do not introduce a new normative context.

Reviewer 5: ```suggestion
* do not introduce a new normative context.
```

Author: discarding the whole block in favor of [this](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1967/files#r2911518287) and resolving open thread

  * A single FOCUS dataset explicitly identified by `<FOCUS Dataset ID>` (e.g., `CostAndUsage`)

* **FOCUS Column**, whereby use of:  
  * `FOCUS column` keyword represents any FOCUS column  
  * `FOCUS column` keyword with a qualifier represents a qualified subset of FOCUS columns (e.g., `FOCUS column containing numeric values`)  
  * A single FOCUS column explicitly identified by `<FOCUS Column ID>` (e.g., `BilledCost`)

* **Custom Column**, whereby use of:  
  * `Custom column` keyword represents any custom column  
  * `Custom column` keyword with a qualifier represents a qualified subset of custom columns (e.g., `Custom column containing numeric values`)

* **Structural sub-elements within Columns** (objects, keys, key values):  
  *Note: MUST NOT use `object`, `key`, or `value` keywords alone. Always reference them in context, e.g.:*  
  * `Object in Columns containing JsonObjectFormat values`  
  * `Key in Object in FOCUS/Custom column containing JsonObjectFormat values`  
  * `Key value in Object in FOCUS/Custom column containing key-value pairs`

The subject SHOULD be explicit and unambiguous.

> **Exception:** A structural anchor requirement MAY use the ID of the enclosing schema construct as its subject, even if that construct is not otherwise an allowed normative subject (e.g., `<FOCUS Attribute  ID>`). This exception applies only to structural anchor requirements.

Reviewer 6: ```suggestion
**Exception:** A structural anchor requirement MAY use the ID of the enclosing schema construct as its subject, even if that construct is not otherwise an allowed normative subject (e.g., `<FOCUS Attribute ID>`). This exception applies only to structural anchor requirements.
```

Double space between "Attribute" and "ID".

### 4. State, Not Behavior

Normative requirements MUST describe a **verifiable state**, not an operational process or behavior.

Specifically:

* Process-oriented verbs such as *ensure*, *handle*, *support*, or *provide* MUST NOT be used.
* If a requirement refers to actor behavior, it MUST be reformulated as:
  * a constraint on the resulting dataset state, or
  * a constraint on a schema-defined artifact.

### 5. Use of BCP 14 Keywords

* Each normative bullet MUST contain exactly one BCP 14 keyword (MUST, SHOULD, MAY, MUST NOT, SHOULD NOT).
* A bullet containing more than one normative keyword MUST be split.

### 6. Splitting Requirements

A requirement MUST be split into multiple bullets if it:

* combines multiple obligations,
* combines a rule and an exception,
* mixes a definition with a constraint,
* applies different constraints to different subjects.

### 7. Composite Requirements

Composite (parent + nested) requirements MAY be used when strictly controlled and MUST adhere to the following requirements:

* Nested bullets MUST share the same condition if defined by the parent bullet.
* Nested bullets SHOULD NOT introduce a different subject.

> * When nested bullets introduce different subjects, all subjects SHOULD be of the same subject type (e.g., FOCUS * columns).

Reviewer 6: ```suggestion
* When nested bullets introduce different subjects, all subjects SHOULD be of the same subject type (e.g., FOCUS columns).
```

Stray `*` between "FOCUS" and "columns" - likely a markdown artifact from the self-suggestion that introduced this line.

* all nested bullets:
  * apply to that same subject,
  * do not introduce a new normative context.

Nested bullets MUST share the same condition if defined by the parent bullet.
Nested bullets SHOULD NOT introduce a different subject. 

> When nested bullets introduce different subjects, all subjects SHOULD be of the same subject type (e.g., FOCUS columns).

Author: ```suggestion
Composite (parent + nested) requirements MAY be used when strictly controlled and MUST adhere to the following requirements:

* Nested bullets MUST share the same condition if defined by the parent bullet.
* Nested bullets SHOULD NOT introduce a different subject.
* When nested bullets introduce different subjects, all subjects SHOULD be of the same subject type (e.g., FOCUS * columns).
```

### 4. State, Not Behavior

Normative requirements MUST describe a **verifiable state**, not an operational process or behavior.

Specifically:

* Process-oriented verbs such as *ensure*, *handle*, *support*, or *provide* MUST NOT be used.
* If a requirement refers to actor behavior, it MUST be reformulated as:
  * a constraint on the resulting dataset state, or
  * a constraint on a schema-defined artifact.

### 5. Use of BCP 14 Keywords

* Each normative bullet MUST contain exactly one BCP 14 keyword (MUST, SHOULD, MAY, MUST NOT, SHOULD NOT).
* A bullet containing more than one normative keyword MUST be split.

### 6. Splitting Requirements

A requirement MUST be split into multiple bullets if it:

* combines multiple obligations,
* combines a rule and an exception,
* mixes a definition with a constraint,
* applies different constraints to different subjects.

### 7. Composite Requirements

Composite (parent + nested) requirements MAY be used when strictly controlled.

They are allowed only when:

* the parent bullet introduces exactly one normative subject, and
* all nested bullets:
  * apply to that same subject,
  * do not introduce a new normative context.

> 

Author: ```suggestion
```

* **FOCUS Dataset** — the primary dataset concept defined by the FOCUS specification.
* **Dataset Instance** — represents a concrete instantiation of a **FOCUS Dataset Instance**.
* **Dataset Artifact** — represents a physical or delivered form of a **FOCUS Dataset Instance Artifact**.

However, by design decision, the specification adopts the following normative conventions:

* **FOCUS Dataset is used as the canonical normative subject** for dataset-level requirements.
* Normative requirements are intentionally written against **FOCUS Dataset**, even when the constraint applies to:
  * a dataset specification,
  * a dataset instance, or
  * a dataset instance artifact.
* The intended level of application (specification vs. instance vs. artifact) is inferred from context rather than encoded in the normative subject.

This choice is intentional and overrides interpretations based solely on abstraction level.

> To support this approach and avoid ambiguity:
>
> * The **FOCUS Dataset** glossary definition SHOULD be refined so that it is no longer interpreted as purely abstract, but as a normative anchor that may represent the dataset specification and its specific implementations.

> > * Glossary entries for **Dataset Instance** and **Dataset Artifact** SHOULD explicitly reflect that they represent **FOCUS Dataset Instance** and **FOCUS Dataset Instance Artifact**, respectively (important with the planned **Native Dataset** entry in version 1.4).

Author: To support this approach and avoid ambiguity:

* The **FOCUS Dataset** glossary definition SHOULD be refined so that it is no longer interpreted as purely abstract, but as a normative anchor that may represent the dataset specification and its specific implementations.
 * Glossary entries for **Dataset Instance** and **Dataset Artifact** SHOULD explicitly reflect that they represent **FOCUS Dataset Instance** and **FOCUS Dataset Instance Artifact**, respectively (important with the planned **Native Dataset** entry in version 1.4).

---

Additional suggestions for glossary entries:

* **FOCUS column**: Column defined by FOCUS and included in a FOCUS dataset
* **Custom column**: Column not defined by FOCUS, prefixed with `x_` and included in a FOCUS dataset
* **Native column**: Column included in a Native dataset

```suggestion
```

* <DatasetId> SHOULD include <ColumnId> when <Condition>.
```

##### 4.1.3. Technical Requirements: Technical Attributes Conformance

```markdown
* <DatasetId> MUST conform to <TechnicalAttributeId> requirements.
```

> ##### 4.1.3. Business Requirements: Business/Contextual Attributes Conformance

Reviewer 4: 🤖 [AI][Claude] Duplicate section number: This should be "4.1.4" - currently both Technical Requirements (line 253) and this Business Requirements section use "4.1.3".

```suggestion
##### 4.1.4. Business Requirements: Business/Contextual Attributes Conformance
```

```

##### 7.1.4. Technical Requirements: Values and Value Ranges

```markdown
* <ColumnId> MUST be a valid decimal value.
* <ColumnId> MUST be a non-negative decimal value.
```

##### 7.1.5. Technical Requirements: Column-to-Column Relationships

```markdown
* <ColumnId> SHOULD/MUST remain consistent over time for a given <OtherColumnId>.
```

##### 7.1.6. Business & Contextual Requirements: Unit/Denomination

```markdown
* <ColumnId> MUST be denominated in the BillingCurrency.
* <ColumnId> MUST be expressed in the <OtherColumnId>.
```

##### 7.1.7. Business & Contextual Requirements: Uniqueness

```markdown
* <ColumnId> MUST be a unique identifier within <Scope>.
```

##### 7.1.8. Business & Contextual Requirements: Fallback/Substitute Values

```markdown
* <ColumnId> MUST NOT duplicate <OtherColumnId> when <Condition>
```

##### 7.1.9. Business & Contextual Requirements: Relationships Outside the Spec

```markdown
* The sum of <ColumnId> in a given billing period MUST/MAY NOT match the sum of the invoices received for that billing period for a billing account.
```

##### 7.1.10. Business & Contextual Requirements: Cost Validation Rules

```markdown

>   * <CostColumnId> MUST equal the product of <UnitPriceColumnId> and PricingQuantity when <UnitPriceColumnId> is not null and PricingQuantity is not null.

Reviewer 5: ```suggestion
* <CostColumnId> MUST equal the product of <UnitPriceColumnId> and PricingQuantity when <UnitPriceColumnId> is not null and PricingQuantity is not null.
```

   2. **Qualified FOCUS Column Requirements:** Applicable to a subset of FOCUS columns that declare conformance to the Attribute and are identified through a qualifier
   3. **Specific FOCUS Column Requirements:** Applicable to a specific FOCUS column, identified explicitly by Column ID.
3. **Custom Column Attribute Requirements:**
   1. **Global Custom Column Requirements:** Applicable to all Custom columns, regardless of their structure or purpose.
   2. **Qualified Custom Column Requirements:** Applicable to a subset of Custom columns, identified through a qualifier.
4. **Sub-element Attribute Requirements:** Applicable to structural sub-elements within Columns that declare conformance to the Attribute.
   1. **Elements (Objects) in Columns containing JsonObjectFormat values**
   2. **Keys in Objects in Columns containing JsonObjectFormat values**
   3. **Key values in Objects in Columns containing JsonObjectFormat values**
   4. **Keys in Columns containing Key-Value pair format values**
   5. **Key values in Columns containing Key-Value pair format values**

### 5. Ordering of Attribute Requirements within Groups

> > TODO:

Author: ```suggestion
* Within each group of requirements, order individual requirements as follows:
  * Requirements targeting the general subject (e.g., FOCUS column) first
  * Requirements targeting qualified subjects (e.g., FOCUS column containing numeric values) after
  * Within each subject, order by BCP 14 keyword:
    * MUST – an absolute requirement
    * MUST NOT – a prohibition
    * SHOULD – recommended but not mandatory
    * SHOULD NOT – discouraged but not strictly prohibited
    * MAY – optional
```

| FOCUS Dataset  | `<FOCUS Dataset ID>` FOCUS dataset | Specific FOCUS dataset identified by Dataset ID |
| Native dataset | Native dataset | |
| Native column  | Native dataset column | |
| Column         | FOCUS column | Column defined by FOCUS and included in a FOCUS dataset |
| Column         | Custom column | Column not defined by FOCUS and included in a FOCUS dataset |
| Column         | FOCUS/Custom column representing national currency | |
| Column         | FOCUS/Custom column representing virtual currency | |
| Column         | FOCUS/Custom column representing an identifier | |
| Column         | FOCUS/Custom column representing a name | |
| Column         | FOCUS/Custom column representing a product offering that incurred the charge | |
| Column         | FOCUS column with `Category` suffix | |
| Column         | FOCUS/Custom column containing numeric values | |
| Column         | FOCUS/Custom column containing date/time values | |
| Column         | FOCUS/Custom column containing string values | |
| Column         | FOCUS/Custom column containing immutable string values | |
| Column         | FOCUS/Custom column containing not-nullable string values | |

> | Column         | FOCUS/Custom column representing charges to mutable entities?? | |

Reviewer 4: 🤖 [AI][Claude] This row contains "??" which suggests incomplete/uncertain content. Please clarify the intended qualifier or remove this row before merging.

Author: resolving in favor of [this thread](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1967/files#r2910696009)

> | Column         | FOCUS/Custom column representing charges to mutable entities?? | |

Author: ```suggestion
| Column         | FOCUS/Custom column containing immutable string values | |
| Column         | FOCUS/Custom column containing mutable string values | |
| Column         | FOCUS/Custom column containing not-nullable string values | |
```

| Column         | FOCUS/Custom column containing values in JsonObjectFormat format | |
| Column         | FOCUS/Custom column containing values in key-value pair format | |
| Column         | `<FOCUS Dataset ID>.<FOCUS Column ID>` | Specific column included in a FOCUS dataset, identified by Column ID |
| Object/Element | Object/Element in array in FOCUS/Custom column containing JsonObjectFormat values | |
| Object         | Object in FOCUS/Custom column containing JsonObjectFormat values | |
| Key            | Keys in Object in FOCUS/Custom column containing JsonObjectFormat values | |
| Key value      | Key values in Object in FOCUS/Custom column containing JsonObjectFormat values | |
| Key            | Keys in FOCUS/Custom column containing values in key-value pair format | |

> | Key value      | Key values in FOCUS/Custom column containing values in key-value pair format | |

Author: ```suggestion
| Column         | FOCUS/Custom column containing values in key-value pair format | |
| Key            | Keys in FOCUS/Custom column containing values in key-value pair format | |
| Key value      | Key values in FOCUS/Custom column containing values in key-value pair format | |
| Column         | FOCUS column containing values in JsonObjectFormat format | |
| Object         | Object in FOCUS column containing JsonObjectFormat values | |
| Key            | Keys in Object in FOCUS column containing JsonObjectFormat values | |
| Key value      | Key values in Object in FOCUS column containing JsonObjectFormat values | |
| Object         | Object in array in FOCUS column containing JsonObjectFormat values | |
| Column         | Custom column containing JSON object | |
| Object         | Object in Custom column containing JSON object | |
| Key            | Keys in Object in Custom column containing JSON object | |
| Key value      | Key values in Object in Custom column containing JSON object | |
| Object         | Object in array in Custom column containing JSON object | |
| Column         | `<FOCUS Dataset ID>.<FOCUS Column ID>` | Specific column included in a FOCUS dataset, identified by Column ID |
```

0. **Structural Attribute Anchor Requirement:** Introduces the scope of the Attribute and provides a stable parsing entry point; it does not introduce a verifiable constraint.
1. **FOCUS Dataset-level Attribute Requirements:**
   1. **Global FOCUS Dataset Requirements:** Applicable to all FOCUS datasets that declare conformance to the Attribute, regardless of their structure, specific role or context.
   2. **Qualified FOCUS Dataset Requirements:** Applicable to a subset of FOCUS datasets that declare conformance to the Attribute and are identified through a qualifier.
   3. **Specific FOCUS Dataset Requirements:** Applicable to a specific FOCUS dataset, identified explicitly by Dataset ID.
2. **FOCUS Column-level Attribute Requirements:** Applicable to FOCUS columns that declare conformance to the Attribute.
   1. **Global FOCUS Column Requirements:** Applicable to all FOCUS columns that declare conformance to the Attribute, regardless of their structure, specific role or context.

>    2. **Qualified FOCUS Column Requirements:** Applicable to a subset of FOCUS columns that declare conformance to the Attribute and are identified through a qualifier

Reviewer 6: ```suggestion
   2. **Qualified FOCUS Column Requirements:** Applicable to a subset of FOCUS columns that declare conformance to the Attribute and are identified through a qualifier.
```

Missing trailing period. Every other item in this list has one.

   3. **Specific FOCUS Column Requirements:** Applicable to a specific FOCUS column, identified explicitly by Column ID.
3. **FOCUS Column sub-element Attribute Requirements:** Applicable to structural sub-elements within columns that declare conformance to the Attribute.

>    1. **Elements (Objects) in Columns containing JsonObjectFormat values**

Author: ```suggestion
   1. **Objects in Columns containing JsonObjectFormat values**
```

   2. **Qualified Custom Column Requirements:** Applicable to a subset of Custom columns, identified through a qualifier.
4. **Sub-element Attribute Requirements:** Applicable to structural sub-elements within Columns that declare conformance to the Attribute.
   1. **Elements (Objects) in Columns containing JsonObjectFormat values**
   2. **Keys in Objects in Columns containing JsonObjectFormat values**
   3. **Key values in Objects in Columns containing JsonObjectFormat values**
   4. **Keys in Columns containing Key-Value pair format values**

>    5. **Key values in Columns containing Key-Value pair format values**

Author: ```suggestion
3. **FOCUS Column sub-element Attribute Requirements:** Applicable to structural sub-elements within columns that declare conformance to the Attribute.
   1. **Elements (Objects) in Columns containing JsonObjectFormat values**
   2. **Keys in Objects in Columns containing JsonObjectFormat values**
   3. **Key values in Objects in Columns containing JsonObjectFormat values**
   4. **Keys in Columns containing Key-Value pair format values**
   5. **Key values in Columns containing Key-Value pair format values**
4. **Custom Column Attribute Requirements:**
   1. **Global Custom Column Requirements:** Applicable to all Custom columns, regardless of their structure or purpose.
   2. **Qualified Custom Column Requirements:** Applicable to a subset of Custom columns, identified through a qualifier.
```

### 5. Ordering of Attribute Requirements within Groups

* Within each group of requirements, order individual requirements as follows:
  * Requirements targeting the general subject (e.g., FOCUS column) first
  * Requirements targeting qualified subjects (e.g., FOCUS column containing numeric values) after
  * Within each subject, order by BCP 14 keyword:

>     * MUST – an absolute requirement

Reviewer 6: ```suggestion
    * **MUST** – an absolute requirement
```

Bold the BCP keyword to match the Dataset (line 208) and Column (line 329) ordering sections.

>     * MUST NOT – a prohibition

Reviewer 6: ```suggestion
    * **MUST NOT** – a prohibition
```

Same: bold keyword for consistency with Dataset/Column sections.

>     * SHOULD – recommended but not mandatory

Reviewer 6: ```suggestion
    * **SHOULD** – recommended but not mandatory
```

Same: bold keyword for consistency.

>     * SHOULD NOT – discouraged but not strictly prohibited

Reviewer 6: ```suggestion
    * **SHOULD NOT** – discouraged but not strictly prohibited
```

Same: bold keyword for consistency.

>     * MAY – optional

Reviewer 6: ```suggestion
    * **MAY** – optional
```

Same: bold keyword for consistency.

The structural anchor requirement:

* uses the Attribute ID as the subject,
* introduces the scope of the requirements,
* is non-verifiable and non-enforceable,
* exists solely for structural consistency and automated parsing.

Canonical form:

``` markdown
<AttributeId> MUST adhere to the following requirements:
```

This is the only case in which an Attribute ID may appear as a normative subject.

### 3. Normative Subjects in Attribute Requirements

All enforceable normative requirements within an Attribute section MUST target schema-level entities that conform to the Attribute.

The Attribute itself MUST NOT be treated as an enforceable subject beyond the structural anchor requirement.

Each Attribute implicitly assumes one or more intended normative subjects.

These intended subjects define which kinds of schema-level entities the Attribute’s requirements are written for.

The following table lists commonly used intended subjects.

***Note:** This list is non-exhaustive and will be extended over time.*

| Subject (Core) | Distinct Subjects with Qualifiers | Description/Note |
|----------------|-----------------------------------|------------------|

> | Attribute      | <Attribute ID> | Specific Attribute identified by Dataset ID |

Author: ```suggestion
| Attribute      | `<Attribute ID>` | Specific Attribute identified by Dataset ID |
```

| FOCUS Dataset  | FOCUS dataset | |

> | FOCUS Dataset  | <FOCUS Dataset ID> | Specific FOCUS dataset identified by Dataset ID |

Author: ```suggestion
| FOCUS Dataset  | <FOCUS Dataset ID> FOCUS dataset | Specific FOCUS dataset identified by Dataset ID |
```

> | FOCUS Dataset  | <FOCUS Dataset ID> FOCUS dataset | Specific FOCUS dataset identified by Dataset ID |

Author: ```suggestion
| FOCUS Dataset  | `<FOCUS Dataset ID>` FOCUS dataset | Specific FOCUS dataset identified by Dataset ID |
```

| Column         | Custom column | Column not defined by FOCUS and included in a FOCUS dataset |
| Column         | FOCUS/Custom column representing national currency | |
| Column         | FOCUS/Custom column representing virtual currency | |
| Column         | FOCUS/Custom column representing an identifier | |
| Column         | FOCUS/Custom column representing a name | |
| Column         | FOCUS/Custom column representing a product offering that incurred the charge | |
| Column         | FOCUS column with `Category` suffix | |
| Column         | FOCUS/Custom column containing numeric values | |
| Column         | FOCUS/Custom column containing date/time values | |
| Column         | FOCUS/Custom column containing string values | |
| Column         | FOCUS/Custom column containing immutable string values | |
| Column         | FOCUS/Custom column containing not-nullable string values | |
| Column         | FOCUS/Custom column representing charges to mutable entities?? | |
| Column         | FOCUS/Custom column containing values in JsonObjectFormat format | |
| Column         | FOCUS/Custom column containing values in key-value pair format | |

> | Column         | <FOCUS Column ID> | Specific column included in a FOCUS dataset, identified by Column ID |

Author: ```suggestion
| Column         | <FOCUS Dataset ID>.<FOCUS Column ID> | Specific column included in a FOCUS dataset, identified by Column ID |
```

> | Column         | <FOCUS Dataset ID>.<FOCUS Column ID> | Specific column included in a FOCUS dataset, identified by Column ID |

Author: ```suggestion
| Column         | `<FOCUS Dataset ID>.<FOCUS Column ID>` | Specific column included in a FOCUS dataset, identified by Column ID |
```
